tests: Port `jobserver-error` to rmake.rs

jieyouxu opened the pull request:
Part of #121876.

This PR ports `tests/run-make/jobserver-error` to rmake.rs, and is basically #128789 slightly adjusted.

The complexity involved here is mostly how to get `/dev/null/` piping to fd 3 working with std `Command`, whereas with a shell this is much easier (as is evident with the `Makefile` version).

Supersedes #128789.
This PR is co-authored with @Oneirical and @coolreader18.

try-job: aarch64-gnu
try-job: i686-gnu-1
try-job: x86_64-gnu-debug
try-job: x86_64-gnu-llvm-18-1

jieyouxu committed:
run-make-support: add `set_aux_fd` helper

Co-authored-by: Noa <coolreader18@gmail.com>
Co-authored-by: Oneirical <manchot@videotron.ca>

jieyouxu committed:
tests: port `jobserver-error.rs` to rmake.rs

Co-authored-by: Noa <coolreader18@gmail.com>
Co-authored-by: Oneirical <manchot@videotron.ca>

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit aaddf77ceec2628a4b69cb35dd58b0b95b506ae0 with merge 2b44a3d7972abebf4ca78446aa89debf512d411e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"aaddf77ceec2628a4b69cb35dd58b0b95b506ae0","merge_sha":"2b44a3d7972abebf4ca78446aa89debf512d411e"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12757836189/job/35558924194) [(plain)](https://github.com/rust-lang/rust/commit/aaddf77ceec2628a4b69cb35dd58b0b95b506ae0/checks/35558924194/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit aaddf77ceec2628a4b69cb35dd58b0b95b506ae0 with merge 163ddb169c67425e679c412d12576fae3f34df92...
<!-- homu: {"type":"TryBuildStarted","head_sha":"aaddf77ceec2628a4b69cb35dd58b0b95b506ae0","merge_sha":"163ddb169c67425e679c412d12576fae3f34df92"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12758746801/job/35565604321)
Build commit: 163ddb169c67425e679c412d12576fae3f34df92 (`163ddb169c67425e679c412d12576fae3f34df92`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12758746801/job/35565604321"},"merge_sha":"163ddb169c67425e679c412d12576fae3f34df92"} -->

jieyouxu commented:
r? compiler

Noratrieb commented on tests/run-make/jobserver-error/rmake.rs:
```diff
@@ -0,0 +1,55 @@
+fn main() {
+    let out = rustc()
+        .stdin_buf(("fn main() {}").as_bytes())
+        .env("MAKEFLAGS", "--jobserver-auth=5,5")
```
```suggestion
        .env("MAKEFLAGS", "--jobserver-auth=1000,1000")
```
use some higher FDs to guard against rustc using more fds initially

jieyouxu replied:
Switched to higher FDs.

Noratrieb commented on src/tools/run-make-support/src/command.rs:
```diff
@@ -151,6 +151,53 @@ impl Command {
+            // SAFETY(io-safety): `fd` is already owned.
+            cvt(unsafe { libc::fcntl(fd.as_raw_fd(), libc::F_SETFD, 0) })
+                .expect("disabling CLOEXEC failed");
+            // The `pre_exec` function should be unconditionally set, since it captures `fd`, and
+            // this ensures that it stays open until the fork.
```
Wouldn't it still be closed once pre_exec returns? I think you need to `mem::forget` it explicitly.
Personally I would put `fd` in a `ManuallyDrop` at the beginning of the function and only work with its `RawFd` from that point on, that seems less prone to mistakes.

jieyouxu replied:
Switched to a `ManuallyDrop`, and used its `RawFd` afterwards.

jieyouxu commented:
Changes since last review:

- Switched to `ManuallyDrop` and use `RawFd` only afterwards.
- Bumped to higher invalid FD and reblessed the stderr snapshot.

@rustbot ready

Noratrieb commented on src/tools/run-make-support/src/command.rs:
```diff
@@ -151,6 +151,53 @@ impl Command {
+            // If the new file descriptor is already the same as fd, just turn off `FD_CLOEXEC`.
+
+            // SAFETY(io-safety): `fd` is already owned.
+            cvt(unsafe { libc::fcntl(fd, libc::F_SETFD, 0) }).expect("disabling CLOEXEC failed");
```
nit: it's more future compatible to first `F_GETFD`, unset the bit, and throw that in there instead of 0

jieyouxu replied:
Changed this to first retrieve the fd flags, then clear the `FD_CLOEXEC` bit, instead of YOLOing a `0` when `F_SETFD`.

Noratrieb commented on src/tools/run-make-support/src/command.rs:
```diff
@@ -151,6 +151,53 @@ impl Command {
+            if fd.as_raw_fd() != new_fd {
+                // SAFETY(io-safety): `new_fd` is not necessarily an unused fd. However, we're
+                // ensuring that `new_fd` will now refer to the same file descriptor as `fd`, which
+                // is safe as long as we manage the lifecycle of both descriptors correctly. This
```
nit: This here should say that it's the caller's responsibility that we won't override the target fd

jieyouxu replied:
Reworded as suggested.

Noratrieb commented on src/tools/run-make-support/src/command.rs:
```diff
@@ -151,6 +151,53 @@ impl Command {
+    ///
+    /// # Safety
+    ///
+    /// Use with caution! Ideally, only set one aux fd; if there are multiple, their old `fd` may
```
nit: I would write this in a way that more clearly states the requirement. though given that the safety requirement is basically impossible to actually uphold (see next comment) I would probably just make this function safe because the unsafe is just noise

jieyouxu replied:
Good point, I just dropped the unsafe.

Noratrieb commented on tests/run-make/jobserver-error/rmake.rs:
```diff
@@ -0,0 +1,55 @@
+        .stderr_utf8();
+    diff().expected_file("cannot_open_fd.stderr").actual_text("actual", out).run();
+
+    // SAFETY(io-safety): we don't have overlapping fd 3.
```
nit: you have no idea (:ferrisClueless:) if this is true or not, std makes no guarantee to not open an arbitrary amount of fds at startup.

but also if it wasn't true it would very likely still be fine anyways

jieyouxu replied:
Indeed, I just dropped the unsafe because it's not really enforceable.

Noratrieb reviewed (approved):
some nits around the safety of the whole situation but you don't have to fix it you can also just r=me and :shipit:

jieyouxu commented:
Addressed the nits. I also deleted the commented-out code in favor the original Makefile bash line to be less confusing.
@rustbot ready

jieyouxu commented:
If a future reader is able to improve the safety/soundness situation of this test, please go right ahead 😅

Noratrieb reviewed (approved):
i think the safety is perfevrly fine

Noratrieb commented:
time for some CI bouncing
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit b59576439f6486a9e85c1c18ede5f86cac3984cb has been approved by `Noratrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Noratrieb b59576439f6486a9e85c1c18ede5f86cac3984cb -->
<!-- homu: {"type":"Approved","sha":"b59576439f6486a9e85c1c18ede5f86cac3984cb","approver":"Noratrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit b59576439f6486a9e85c1c18ede5f86cac3984cb with merge ea1fd7a513204425688ae747008b5705d4a1f2c7...
<!-- homu: {"type":"BuildStarted","head_sha":"b59576439f6486a9e85c1c18ede5f86cac3984cb","merge_sha":"ea1fd7a513204425688ae747008b5705d4a1f2c7"} -->

rust-log-analyzer commented:
The job **`i686-mingw`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12891505882/job/35943657202) [(plain)](https://github.com/rust-lang-ci/rust/commit/ea1fd7a513204425688ae747008b5705d4a1f2c7/checks/35943657202/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling run_make_support v0.2.0 (C:\a\rust\rust\src\tools\run-make-support)
error: unused import: `mem`
 --> src\tools\run-make-support\src\command.rs:5:16
  |
5 | use std::{ffi, mem, panic};
  |
  = note: `-D unused-imports` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unused_imports)]`


[RUSTC-TIMING] run_make_support test:false 0.938
error: could not compile `run_make_support` (lib) due to 1 previous error
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 1:28:25
make: *** [Makefile:120: ci-mingw-x] Error 1
  network time: Tue, 21 Jan 2025 18:09:41 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12891505882/job/35948972379)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/12891505882/job/35948972379","builder_name":"checks-actions"} -->

Noratrieb commented:
hiii bors

jieyouxu commented:
Rescoped the import to the `cfg(unix)`'d code path.
@bors r=@Noratrieb

bors commented:
:pushpin: Commit 8f00fbe461a37d3c7bf04e6c0e4b086721c3df98 has been approved by `Noratrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Noratrieb 8f00fbe461a37d3c7bf04e6c0e4b086721c3df98 -->
<!-- homu: {"type":"Approved","sha":"8f00fbe461a37d3c7bf04e6c0e4b086721c3df98","approver":"Noratrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout migrate-jobserver-errors` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self migrate-jobserver-errors --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Removing tests/run-make/jobserver-error/Makefile
Auto-merging src/tools/tidy/src/allowed_run_make_makefiles.txt
CONFLICT (content): Merge conflict in src/tools/tidy/src/allowed_run_make_makefiles.txt
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

jieyouxu commented:
Merge conflicted due to allow-list.
@bors r=Noratrieb

bors commented:
:pushpin: Commit 20229200c80c7a6400f455a2cc2281045c302ba6 has been approved by `Noratrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Noratrieb 20229200c80c7a6400f455a2cc2281045c302ba6 -->
<!-- homu: {"type":"Approved","sha":"20229200c80c7a6400f455a2cc2281045c302ba6","approver":"Noratrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 20229200c80c7a6400f455a2cc2281045c302ba6 with merge cf577f34c47937ccb9983186eca5f8719da585f4...
<!-- homu: {"type":"BuildStarted","head_sha":"20229200c80c7a6400f455a2cc2281045c302ba6","merge_sha":"cf577f34c47937ccb9983186eca5f8719da585f4"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12922933737/job/36044913799)
Approved by: Noratrieb
Pushing cf577f34c47937ccb9983186eca5f8719da585f4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Noratrieb","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12922933737/job/36044913799"},"merge_sha":"cf577f34c47937ccb9983186eca5f8719da585f4"} -->

rust-timer commented:
Finished benchmarking commit (cf577f34c47937ccb9983186eca5f8719da585f4): [comparison URL](https://perf.rust-lang.org/compare.html?start=3cd8fcbf87bd28a1f31be000ca906fb66f4d451d&end=cf577f34c47937ccb9983186eca5f8719da585f4&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=3cd8fcbf87bd28a1f31be000ca906fb66f4d451d&end=cf577f34c47937ccb9983186eca5f8719da585f4&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3cd8fcbf87bd28a1f31be000ca906fb66f4d451d&end=cf577f34c47937ccb9983186eca5f8719da585f4&stat=max-rss)
<details>
<summary>Results (secondary -1.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.8% | [-4.1%, -0.6%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=3cd8fcbf87bd28a1f31be000ca906fb66f4d451d&end=cf577f34c47937ccb9983186eca5f8719da585f4&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=3cd8fcbf87bd28a1f31be000ca906fb66f4d451d&end=cf577f34c47937ccb9983186eca5f8719da585f4&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 776.393s -> 776.757s (0.05%)
**Artifact size**: 325.90 MiB -> 325.99 MiB (0.03%)
